src/System.Resources.Extensions/src/System/Resources/Extensions/TypeNameComparer.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System.Resources.Extensions
+{
+    /// <summary>
+    /// Compares type names as strings, ignoring version.
+    /// When type names are missing, mscorlib is assumed.
+    /// This comparer is not meant to capture all scenarios (eg: TypeForwards)
+    /// but is meant to serve as a best effort, avoiding false positives, in the 
+    /// absense of real type meatadata.
+    /// </summary>
+    internal class TypeNameComparer : IEqualityComparer<string>
+    {
+        public static TypeNameComparer Instance { get; } = new TypeNameComparer();
+
+        private static readonly char[] s_whiteSpaceChars =
+        {
+            ' ', '\n', '\r', '\t'
+        };
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I think our policy generally is to not force inlining unless we have proof that it is not inlining automatically and that it causing a real world performance problem - or something like that. Surely this would inline without trouble.

Reply:
I think this won't be inlined because of the method's IL size. (I believe it 26 bytes?)

Reply:
So why do we think we know better (not trying to be facetious)

Reply:
Our general policy is that any use of AggressiveInlining should be accompanied by perf numbers, as there needs to be a big enough win to counteract the potential negatives that come from the increased code size.

Reply:
So the most common case this code will be hit and it does a significant amount is comparing is dealing with equal primitive names since those are the only ones that will have a hash collision.

I measured this with in-lining various methods.
1. All
2. ReadTypeName, ReadSimpleName
3. ReadSimpleName
4. None

|  Method |         Type1 |                Type2 |     Mean |    Error |   StdDev | Rank |
|-------- |-------------- |--------------------- |---------:|---------:|---------:|-----:|
| Equals1 | System.String | Syste(...)4e089 [90] | 82.24 ns | 1.480 ns | 1.235 ns |    1 |
| Equals2 | System.String | Syste(...)4e089 [90] | 82.45 ns | 1.232 ns | 1.092 ns |    1 |
| Equals3 | System.String | Syste(...)4e089 [90] | 86.39 ns | 1.664 ns | 1.635 ns |    2 |
| Equals4 | System.String | Syste(...)4e089 [90] | 89.66 ns | 1.674 ns | 1.566 ns |    3 |

So in-lining both the Read methods is 9% savings.  Looks like I can remove the in-lining of IsMscorlib since JIT was likely already doing this.  @stephentoub how should I reason about this tradeoff?  Note these methods are private and only called twice each.

Reply:
> only called twice each.

Do those callers get inlined when these get inlined?  If not, then inlining these should be fine and worth the boost, since you're then only ending up with a few additional instructions.  If those callers do get inlined, then you'd want to understand how many copies of this code will actually show up.  It's a bigger concern when the methods being inlined may have an unbounded number of call sites (e.g. because they're public), show up in generics (and thus could have a multiplicative effect with generic instantations), etc.

Reply:
Got it.  The callers are too big to be inlined and are not generic.  I understand the reasoning you describe here.  Appreciate your help.